specification/under_development/profiles/Base_Profile.md

```diff
@@ -0,0 +1,484 @@
+  the entry point
+- An attribute named `"required_results"` indicating the maximal number of
+  measurement results that need to be stored while executing the entry point
+  function
+
+Optionally, additional attributes may be attached to the entry point. Any custom
+function attributes attached to the entry point will be reflected as metadata in
+the program output; this includes both mandatory and optional attributes but not
+parameter attributes or return value attributes. This in particular implies that
+the [labeling schema](#output-recording) used in the recorded output can be
+identified by looking at the metadata in the produced output. See the
+specification of the [output schemas](../output_schemas/) for more information
+about how metadata is represented in the output schema.
+
+Custom function attributes will show up as part of an [attribute
+group](https://releases.llvm.org/13.0.1/docs/LangRef.html#attrgrp) in the IR.
+Attribute groups are numbered such that they can be easily referenced by
+multiple function definitions or global variables. Consumers of Base Profile
+compliant programs must not rely on a particular numbering, but instead look for
+function to which an attribute with the name `"entry_point"` is attached to
+determine which one to invoke when the program is launched.
+
+Within the restrictions imposed by the Base Profile, the number of qubits that
+are needed to execute the quantum program must be known at compile time. This
+number is captured in the form of the `"required_qubits"` attribute attached to
+the entry point. The value of the attribute must be the string representation of
+a 64-bit integer constant.
+
+Similarly, the number of measurement results that need to be stored when
+executing the entry point function is captured by the `"required_results"`
+attribute. Since qubits cannot be used after measurement, in the case of the
+Base Profile, this value is equal to the number of measurement results in the
+program output.
+
+For a program to be valid, both the number of qubits used and the number of
+measurements to store must be strictly positive. More details on the usage of
+[qubits and results](#qubit-and-result-usage) in general can be found in the
+sections below.
+
+Beyond the entry point specific requirements related to attributes, custom
+attributes may optionally be attached to any of the declared functions.
+Furthermore, the following [LLVM
+attributes](https://llvm.org/docs/LangRef.html#function-attributes) may be used
+according to their intended purpose on function declarations and call sites:
+`inlinehint`, `nofree`, `norecurse`, `readnone`, `readonly`, `writeonly`, and
+`argmemonly`.
+
+### Function Body
+
+The function body consists of two blocks, connected by an unconditional
+branching into the second block at the end of the first block.
+
+The first block contains only calls to QIS functions. Any number of calls to QIS
```

Comment:
A bit ambiguous. It could contain other things, such as any LLVM computations along these calls to QIS functions. As said below, the limit between the two blocks is rather when calls to output functions are first made.

Reply:
Yes, it is *really* only calls (aside from an unconditional branch at the end of the block). It is important to note that any constant folding and compile time evaluation that can happen does happen as part of compiling to a profile. The only thing that remains in the IR is anything that is truely only know at runtime (which really is only the result of a measurement or any expression that in some form or another depends on the result of a measurement). Since the Base Profile does not permit to do anything to a qubit after measuring it, it also means that all classical expressions will be evaluated at compile time.